src/main/java/org/jabref/gui/exporter/SaveDatabaseAction.java

```diff
@@ -216,48 +220,45 @@ private boolean save(Path targetPath, SaveDatabaseMode mode) {
     }
 
     private boolean saveDatabase(Path file, boolean selectedOnly, Charset encoding, SavePreferences.DatabaseSaveType saveType) throws SaveException {
+
+
         SavePreferences preferences = this.preferences.getSavePreferences()
                                                       .withEncoding(encoding)
                                                       .withSaveType(saveType);
-        try (AtomicFileWriter fileWriter = new AtomicFileWriter(file, preferences.getEncoding(), preferences.shouldMakeBackup())) {
-            BibtexDatabaseWriter databaseWriter = new BibtexDatabaseWriter(fileWriter, preferences, entryTypesManager);
 
-            if (selectedOnly) {
-                databaseWriter.savePartOfDatabase(panel.getBibDatabaseContext(), panel.getSelectedEntries());
-            } else {
-                databaseWriter.saveDatabase(panel.getBibDatabaseContext());
-            }
-
-            panel.registerUndoableChanges(databaseWriter.getSaveActionsFieldChanges());
+        Consumer<List<FieldChange>> consumer = fieldChanges -> panel.registerUndoableChanges(fieldChanges);
 
-            if (fileWriter.hasEncodingProblems()) {
-                saveWithDifferentEncoding(file, selectedOnly, preferences.getEncoding(), fileWriter.getEncodingProblems(), saveType);
+        var future = saveManager.save(file, selectedOnly, encoding, saveType, consumer);
+        Set<Character> characters;
+        try {
+            characters = future.get();
+            if (!characters.isEmpty()) {
```

Comment:
it's not really understandable what's going on here (especially in contrast to the easy "hasEncodingProblems" check that was used before

Reply:
I offloaded the saving to the SaveManager which is the actual goal of this whole process.
The actual saving writing is offloaded to the throttler in the GlobalSaveManager.
However, I still have to check for encoding problems and as they solely are dependig on the file writer which is iside a try clause I needed a way to return the data.
And the saveWithDifferentEncoding method is a gui method which shows a dialog the user.

Reply:
I agree! What I meant is that the current code is not really understandable, since it's not clear what characters are returned. I first thought that these are the characters written to the file... 

That's why I proposed below to have a simple class that is returned with understandable methods (like `getEncodingProblems`), similar to the ParsingResult class.